enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,407 @@
+  * These pods will be scheduled across kubelet's "defaultCpuSet"
+  * "defaultCpuSet" = CPUs not used by the platform or pods with guaranteed CPUs
+  * This is a dynamic cpuset that will re-size based the the creation/deletion of pods with guaranteed CPUs
+* Workload 3 - Application user plane pods – L1/L2 function
+  * These are the pods that carry the user plane traffic
+  * These pods will use guaranteed CPUs along with additional isolation for example:
+    * Disable CPU load balancing and CFS quota
+    * Isolate interrupts from the allocated CPUs
+    * Reduce timer ticks (nohz_full)
+  * Generally a small number of pods consuming the bulk of the CPUs on the node
+
+Although there are effectively two CPU pools from an application perspective, Kubernetes only knows about a 
+single pool of CPUs (apart from the reserved CPU pool) and uses this pool for all containers. Kubernetes 
+dynamically allocates CPUs from this single pool for guaranteed QoS containers (using whole CPUs) and 
+updates the CPU affinity of the other containers to ensure they do not run on the guaranteed CPUs.
+
+In order to meet the stringent latency, jitter and performance targets of workload 3, additional kernel 
+tuning is required:
+* Some kernel tuning can be done dynamically and applied at guaranteed container creation time.
+* Unfortunately, some kernel tuning needs to be defined at boot time and is static meaning the same tuning is
+applied to all application CPUs whether it is required or not.
+* Some of these static tunings are problematic for workload 2, such as nohz_full and rcu callbacks. So we
+need the tuning for workload 3 but it is problematic for workload 2.
+
+This proposal provides an option for creating a separate pre-defined pool of CPUs for workload 2 (i.e. 
+shared CPUs) and workload 3 (i.e guaranteed CPUs), which will allow the necessary kernel tuning to be 
+applied to workload 3 without adversely affecting workload 2.
+
+### User Stories
+
+* As a telco service provider, I want to configure the set of CPUs running the L1/L2 functions of a vDU 
+application for extremely low latency, without impacting the other application components running on the 
+same server (on different CPUs).
+
+### Goals
+
+* Allow the configuration of separate pre-defined pools of shared and guaranteed CPUs, with different
+kernel tuning applied to each pool.
+* Automatically assign containers that use whole CPUs and are in a pod with the guaranteed QoS class to the
+guaranteed CPUs.
+* The selection of shared vs. guaranteed CPUs should be transparent to the user - no changes to the pod
+spec should be required.
+* The configuration should be optional and activated with a feature gate.
+* Although the initial usecase for this enhancement is single-node deployments, it should also be usable
+for multi-node deployments.
+
+### Non-Goals
+
+* This enhancement assumes the feature is activated as part of installing the cluster and cannot be
```

Comment:
The feature should be activated only when `spec.cpu.shared` specified in PerformanceProfile isn't it?

Reply:
So what part should be activated during installation and what part by specifying `spec.cpu.shared` in the profile?
I mean if user want to edit its profile and add `spec.cpu.shared` is it possible?

Reply:
The whole feature will need to be activated at install time (both the admission hook and the changes to kubelet). So spec.cpu.shared needs to be specified at installation time. It should be possible to changed the number of shared/guaranteed CPUs after installation, but not remove them (or add them on an already running system). The issue is that if you disable the feature, any existing pods with the new shared/guaranteed extended resources could not be scheduled. Likewise, if you tried to turn on the feature, any existing pods would not have the new shared/guaranteed extended resources, so they would be scheduled on nodes without doing the accounting for those resources and we would end up oversubscribing the shared or guaranteed CPUs on those nodes.

Maybe there is some way to address that?

Reply:
For the scheduling issue I think we can use the approach I described here and in the following comments:
https://github.com/openshift/enhancements/pull/1421#discussion_r1228388054 

But for making the feature dynamic you should be able to handle existing pods that are already running in the cluster, or pods that are not going through the scheduler (pods with `nodeName` or static pods) as well.
 
 Since the activation of the feature requires reboot, all the pods would go through the admission phase handled by Kubelet. Kubelet has all the information about running pods, so it can calculates for which pool each pod needs to be assign and set the CPU affinity respectively (similar to what CPU manager is doing today when `reservedSystemCPUs` value changes).

Now, if you're worried about oversubscribing to certain pool by the existing resources, we can use automated tools like PerformaceProfileCreator or others to compute the size of the pools for the user.

Reply:
I added a note here to indicate that the limitation of activating the feature only at installation time could be lifted depending on what the final design looks like.